New Qoqo Gate Guide (#596)
HQSquantumsimulations/qoqo: add_new_gate.md

# GUIDE: HOW TO ADD A NEW GATE TO ROQOQO/QOQO

Depending on its type, adding support for a new quantum gate to roqoqo/qoqo can be more or less straightforward. As a rule of thumb, looking at the already available implementations of similar gates should help in understanding what to do.

All supported gates are currently defined in the `roqoqo/src/operations` module (for qoqo, `qoqo/src/operations`). The specific script to put a new definition in depends on the type of gate.

The general idea is: a `pub struct` defines the name, qubit-related fields as well as parameters (if any). Then, in order to add any kind of property to the gate, `traits` are either derived or manually implemented for it.

> As an examples, we'll take the roqoqo and qoqo definition of `RotateX`.

Reviewer: ```suggestion
As an example, we'll take the roqoqo and qoqo definition of `RotateX`.
```

    ///
    /// # Returns
    ///
    /// * `beta_i` - The imaginary part Im(β) of the off-diagonal elements of the single-qubit unitary matrix.
    fn beta_i(&self) -> CalculatorFloat {
        (self.theta.clone() / 2.0).sin() * (-1.0)
    }
    /// Returns global_phase parameter of the operation.
    ///
    /// # Returns
    ///
    /// * `global_phase` - The global phase φ of the single-qubit unitary.
    fn global_phase(&self) -> CalculatorFloat {
        CalculatorFloat::from(0.0)
    }
}
```

In order, we have:
* Generic text definition of the gate
    ```rust
    /// The XPower gate exp(-i * θ/2 * σ^x).
    ///
    ```
* List of derive instructions, meaning traits that will be automatically implemented for the gate
    ```rust
    #[derive(
        Debug,
        Clone,
        PartialEq,
        roqoqo_derive::InvolveQubits,
        roqoqo_derive::SupportedVersion,
        roqoqo_derive::Operate,
        roqoqo_derive::Substitute,
        roqoqo_derive::OperateSingleQubit,
        roqoqo_derive::Rotate,
    )]
    #[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
    #[cfg_attr(feature = "json_schema", derive(schemars::JsonSchema))]
    ```

>     Traits represent "properties" one can add to a gate. The idea, after defining the gate's `struct`, is to add these already defined properties to the gate.

Reviewer: ```suggestion
    Traits represent "properties" one can add to a gate. The idea, after defining the gate's `struct`, is to add these previously defined properties to the gate.
```

>     Later on the guide will explain the meaning of all of these.

Reviewer: ```suggestion
    Later on, the guide will explain the meaning of all of these.
```

* Struct containing definition of the gate
    ```rust
    pub struct RotateX {
        /// The qubit the unitary gate is applied to.
        qubit: usize,
        /// The angle θ of the rotation, in the interval from 0 to 2 * 2π.
        theta: CalculatorFloat,
    }
    ```

    In general, structs define the signature of the gate: the name, the fields related to referencing qubits as well as the gate's parameter(s) (if any). In order to make the derivation of `traits` work, some of those fields are reserved, specifically:
    ```rust
    /// Array of field names that are reserved for use with specific traits
    const RESERVED_FIELDS: &[&str; 15] = &[
        "qubit",
        "control",
        "control_0",

>         "control_1",

Reviewer: Missing control_2

        "target",
        "qubits",
        "global_phase",
        "alpha_r",
        "alpha_i",
        "beta_r",
        "beta_i",
        "name",
        "mode",
        "mode_0",
        "mode_1",
    ];
    ```
    ##### In `roqoqo-derive/src/lib/rs`. If necessary, more can be added.

>     In this example, the `qubit` one. As for another example, a new two-qubit gate that is characterized by a `target` and `control` qubits will have to have those two fields (always of `usize` type) and the derivation of traits will work as expected, without having to do it manually.

Reviewer: ```suggestion
    In this example, we use the `qubit` one. As for another example, a new two-qubit gate that is characterized by a `target` and `control` qubits will have to have those two fields (always of `usize` type) and the derivation of traits will work as expected, without having to do it manually.
```

* Tags, associating the new gate to different granularities of gate types
    ```rust
    #[allow(non_upper_case_globals)]
    const TAGS_RotateX: &[&str; 5] = &[
        "Operation",
        "GateOperation",
        "SingleQubitGateOperation",
        "Rotation",
        "RotateX",
    ];
    ```

>     These are related to the traits. In the example, "Operation" indicates that is a gate that implements the `Operate` trait; "GateOperation" the `OperateGate` trait; "SingleQubitGateOperation" the `OperateSingleQubitGate` trait and "Rotation" the `Rotate` trait. The last has to always be the gate's name. 

Reviewer: ```suggestion
    These are related to the traits. In the example, "Operation" indicates that is a gate that implements the `Operate` trait; "GateOperation" the `OperateGate` trait; "SingleQubitGateOperation" the `OperateSingleQubitGate` trait and "Rotation" the `Rotate` trait. The last tag must always be the gate's name. 
```

        ///
        /// # Returns
        ///
        /// * `alpha_i` - The imaginary part Im(α) of the on-diagonal elements of the single-qubit unitary matrix.
        fn alpha_i(&self) -> CalculatorFloat {
            CalculatorFloat::from(0.0)
        }
        /// Returns the beta_r parameter of the operation.
        ///
        /// # Returns
        ///
        /// * `beta_r` - The real part Re(β) of the off-diagonal elements of the single-qubit unitary matrix.
        fn beta_r(&self) -> CalculatorFloat {
            CalculatorFloat::from(0.0)
        }
        /// Returns the beta_i parameter of the operation.
        ///
        /// # Returns
        ///
        /// * `beta_i` - The imaginary part Im(β) of the off-diagonal elements of the single-qubit unitary matrix.
        fn beta_i(&self) -> CalculatorFloat {
            (self.theta.clone() / 2.0).sin() * (-1.0)
        }
        /// Returns global_phase parameter of the operation.
        ///
        /// # Returns
        ///
        /// * `global_phase` - The global phase φ of the single-qubit unitary.
        fn global_phase(&self) -> CalculatorFloat {
            CalculatorFloat::from(0.0)
        }
    }
    ```

>     These are gate-specific properties that cannot be derived. In this case, `OperateGate` and `OperateSingleQubitGate` are related to the matrix representation of the gate, so something that has be to manually defined.

Reviewer: ```suggestion
    These are gate-specific properties that cannot be derived. In this case, `OperateGate` and `OperateSingleQubitGate` are related to the matrix representation of the gate, and are therefore something that has be to manually defined.
```
